core/src/main/java/org/apache/iceberg/SnapshotParser.java

```diff
@@ -146,19 +151,31 @@ static Snapshot fromJson(FileIO io, JsonNode node) {
     }
 
     Integer schemaId = JsonUtil.getIntOrNull(SCHEMA_ID, node);
+    String keyMetadataLocation = JsonUtil.getStringOrNull(KEY_METADATA_LOCATION, node);
 
     if (node.has(MANIFEST_LIST)) {
       // the manifest list is stored in a manifest list file
       String manifestList = JsonUtil.getString(MANIFEST_LIST, node);
-      return new BaseSnapshot(
-          io, sequenceNumber, snapshotId, parentId, timestamp, operation, summary, schemaId, manifestList);
+      return new BaseSnapshot(io, sequenceNumber, snapshotId, parentId, timestamp, operation, summary, schemaId,
+          manifestList, keyMetadataLocation);
 
     } else {
       // fall back to an embedded manifest list. pass in the manifest's InputFile so length can be
       // loaded lazily, if it is needed
       List<ManifestFile> manifests = Lists.transform(JsonUtil.getStringList(MANIFESTS, node),
           location -> new GenericManifestFile(io.newInputFile(location), 0));
-      return new BaseSnapshot(io, snapshotId, parentId, timestamp, operation, summary, schemaId, manifests);
+
+      ByteBuffer keyMetadata = null;
+      if (keyMetadataLocation != null) {
+        try (InputStream stream = io.newInputFile(keyMetadataLocation).newStream()) {
+          keyMetadata = ByteBuffer.wrap(ByteStreams.toByteArray(stream));
+        } catch (IOException e) {
+          throw new UncheckedIOException("Failed to read key metadata at location " + keyMetadataLocation, e);
+        }
+      }
```

Comment:
Nit: This seems to be repeated. Would it make sense to put this into a `EncryptionUtils` or `EncKeyUtils` or similar?

Might be premature as I'm not sure of all of the possible cases and I don't _love_ when an overly generic utility class is introduced that has the potential to get relatively large. Though that might not be the case here.

Reply:
I agree, this seems like something we could put in an IO utility.

Also, do we expect key metadata to be reused across snapshots? If so, then we should add some indirection to cache key metadata by location.